examples/ensemble/plot_feature_transformation.py

```diff
@@ -3,118 +3,189 @@
-fpr_rf_lm, tpr_rf_lm, _ = roc_curve(y_test, y_pred_rf_lm)
-
-# Supervised transformation based on gradient boosted trees
-grd = GradientBoostingClassifier(n_estimators=n_estimator)
-grd_enc = OneHotEncoder()
-grd_lm = LogisticRegression(max_iter=1000)
-grd.fit(X_train, y_train)
-grd_enc.fit(grd.apply(X_train)[:, :, 0])
-grd_lm.fit(grd_enc.transform(grd.apply(X_train_lr)[:, :, 0]), y_train_lr)
-
-y_pred_grd_lm = grd_lm.predict_proba(
-    grd_enc.transform(grd.apply(X_test)[:, :, 0]))[:, 1]
-fpr_grd_lm, tpr_grd_lm, _ = roc_curve(y_test, y_pred_grd_lm)
-
-# The gradient boosted model by itself
-y_pred_grd = grd.predict_proba(X_test)[:, 1]
-fpr_grd, tpr_grd, _ = roc_curve(y_test, y_pred_grd)
-
-# The random forest model by itself
-y_pred_rf = rf.predict_proba(X_test)[:, 1]
-fpr_rf, tpr_rf, _ = roc_curve(y_test, y_pred_rf)
-
-plt.figure(1)
-plt.plot([0, 1], [0, 1], 'k--')
-plt.plot(fpr_rt_lm, tpr_rt_lm, label='RT + LR')
-plt.plot(fpr_rf, tpr_rf, label='RF')
-plt.plot(fpr_rf_lm, tpr_rf_lm, label='RF + LR')
-plt.plot(fpr_grd, tpr_grd, label='GBT')
-plt.plot(fpr_grd_lm, tpr_grd_lm, label='GBT + LR')
-plt.xlabel('False positive rate')
-plt.ylabel('True positive rate')
-plt.title('ROC curve')
-plt.legend(loc='best')
-plt.show()
-
-plt.figure(2)
-plt.xlim(0, 0.2)
-plt.ylim(0.8, 1)
-plt.plot([0, 1], [0, 1], 'k--')
-plt.plot(fpr_rt_lm, tpr_rt_lm, label='RT + LR')
-plt.plot(fpr_rf, tpr_rf, label='RF')
-plt.plot(fpr_rf_lm, tpr_rf_lm, label='RF + LR')
-plt.plot(fpr_grd, tpr_grd, label='GBT')
-plt.plot(fpr_grd_lm, tpr_grd_lm, label='GBT + LR')
-plt.xlabel('False positive rate')
-plt.ylabel('True positive rate')
-plt.title('ROC curve (zoomed in at top left)')
-plt.legend(loc='best')
-plt.show()
+rt_model = make_pipeline(
+    random_tree_embedding, LogisticRegression(max_iter=1000)
+)
+rt_model.fit(X_train_linear, y_train_linear)
+
+# %%
+# For the random forest and gradient boosting trees, we need to first find
+# the id of the leaves which will be used as categories.
+
+
+# %%
+def find_tree_leaves(tree):
+    return [
+        node_id
+        for node_id in range(tree.tree_.node_count)
+        if tree.tree_.children_left[node_id] == -1
+    ]
+
+
+# %%
+# Then, we can pipeline random forest or gradient boosting with a logistic
+# regression. However, the feature transformation will happen by calling the
+# method `apply`. The pipeline in scikit-learn expects a call to `transform`.
+# Therefore, we wrapped the call to `apply` within a `FunctionTransformer`.
+
+# %%
+from sklearn.preprocessing import FunctionTransformer
+from sklearn.preprocessing import OneHotEncoder
+
+
+def rf_apply(X, model):
+    return model.apply(X)
+
+
+rf_leaves_yielder = FunctionTransformer(
+    rf_apply, kw_args={"model": random_forest}
+)
+rf_categories = [find_tree_leaves(tree) for tree in random_forest.estimators_]
+
+rf_model = make_pipeline(
+    rf_leaves_yielder,
+    OneHotEncoder(categories=rf_categories),
```

Comment:
Should we use `handle_unknown='ignore'` here to avoid the complication of having to define `categories`?

This way the example is focused on the tree embedding part.